datalad/support/annexrepo.py

```diff
@@ -3327,12 +3328,65 @@ def _save_add(self, files, git=None, git_opts=None):
+    def _save_post(self, message, status, partial_commit, amend=False):
+
+        if amend and self.is_managed_branch() and \
+                self.format_commit("%B").strip() == "git-annex adjusted branch":
+            # We must not directly amend on an adjusted branch, but fix it
+            # up after the fact. That is if HEAD is a git-annex commit.
+            # Otherwise we still can amend-commit normally.
+            amend = False
+            adjust_amend = True
+        else:
+            adjust_amend = False
+
         # first do standard GitRepo business
         super(AnnexRepo, self)._save_post(
-            message, status, partial_commit)
+            message, status, partial_commit, amend, allow_empty=adjust_amend)
         # then sync potential managed branches
         self.localsync(managed_only=True)
+        if adjust_amend:
+            # We committed in an adjusted branch, but the goal is to amend in
+            # corresponding branch.
+
+            adjusted_branch = self.get_active_branch()
+            corresponding_branch = self.get_corresponding_branch()
+
+            org_commit_pointer = corresponding_branch + "~1"
+            author_name = self.format_commit("%an", org_commit_pointer)
+            author_email = self.format_commit("%ae", org_commit_pointer)
+            author_date = self.format_commit("%ad", org_commit_pointer)
+
+            new_env = (self._git_runner.env
+                       if self._git_runner.env else os.environ).copy()
+            # `message` might be empty - we need to take it from the to be
+            # amended commit in that case:
+            msg = message or self.format_commit("%B", org_commit_pointer)
+            with make_tempfile(dir=str(self.dot_git),
+                               prefix="datalad-",
+                               suffix=".index") as index_file:
+                new_env.update({
+                    'GIT_INDEX_FILE': index_file,
+                    'GIT_AUTHOR_NAME': author_name,
+                    'GIT_AUTHOR_EMAIL': author_email,
+                    'GIT_AUTHOR_DATE': author_date
+                })
+                self._call_git(["read-tree", corresponding_branch],
+                               env=new_env,
+                               read_only=False)
+                # TODO: What about fake_dates in this context?
+                #       ATM this is applied on top by _call_git().
+                out, _ = self._call_git(["commit-tree",
+                                         corresponding_branch + "^{tree}",
+                                         "-p", corresponding_branch + "~2",
+                                         "-m", msg],
+                                        env=new_env,
+                                        read_only=False)
+                new_sha = out.strip()
+
+            self.update_ref("refs/heads/" + corresponding_branch, new_sha)
+            self.update_ref("refs/basis/" + adjusted_branch, new_sha)
```

Comment:
I'm just looking at this bit and haven't tried to take in any of the larger change, but do you need to use a temporary index here?  It looks like you already have the tree you want.

<details>
<summary>example</summary>

```sh
set -eu

cd "$(mktemp -d "${TMPDIR:-/tmp}"/g-XXXXXXX)"
git init
git commit --allow-empty -m c1

branch=$(git symbolic-ref HEAD)

echo one >f1
git add f1
git commit -mf1

echo two >f2
git add f2
git commit -mf2

# Go elsewhere to test an out-of-tree amend.
git checkout HEAD~2
git ls-files

git update-ref "$branch" "$(git commit-tree -m amended -p "$branch"~2 "$branch^{tree}")" "$branch"

git log --oneline --stat --decorate master
```

```
[...]
c6b9e14 (master) amended
 f1 | 1 +
 f2 | 1 +
 2 files changed, 2 insertions(+)
293b68e (HEAD) c1
```

</details>

Reply:
No, don't need it here anymore, agree. Which is why it's an indentation less than the lines above. So, one of us is reading it wrong.

Reply:
Sorry, by "here" I didn't mean right here at this line.  Let's try dropping "here": do you need to use a temporary index?

Reply:
Ah, I see. You mean specifying the parent and referencing branch^{tree} should give all that's needed. You're probably right. 
The idea stemmed from thinking about amending in the adjusted branch, where cahnges would be based on a different state. But we don't. It's all committed already.
Will try.

Reply:
Done. Thank you for spotting this, @kyleam !